### DISCUSSION TITLE
Apply OR function between two characters

### DISCUSSION THREAD

**Discussion by agganitk** [board707/DMD_STM32#59] [General]
I want to display language other than english, so able to display it successfully. 
But there is gap between some characters and there is also overlapping. As I far I know I need to adjust Advance and offset parmeters in GFXglyph - I adjusted and got some success.
To appear text correctly , I also need to apply OR function between two characters. Any advice. !!
Do i need to use GRAPHICS_OR .... if yes then how to select, I dont see any function.

**Comment by board707** [comment]
> As I far I know I need to adjust Advance and offset parmeters in GFXglyph - I adjusted and got some success.

Are you talking about a newly created custom font?  What is the language of the font?  All existed library fonts do not needs the adjustments, as far I know.

> To appear text correctly , I also need to apply OR function between two characters.

Please to clarify this. What does it mean?

**Reply by agganitk** [reply]
Yes I am talking about custom font. Language is Hindi.
when there is overlapping between two characters , the result text should be OR of these two character's matrix.

**Comment by board707** [comment]
> Language is Hindi.

A couple months ago I tried to show a Devanagary in my library:
https://youtu.be/Ltrcb_mpjvs

> there is overlapping between two characters , the result text should be OR of these two character's matrix.

Sorry, but OR output is not supported in the library. To show the letters correctly you have to edit glyphs, set all overlapping margins transparent:
https://community.pixelmatix.com/uploads/db4404/original/2X/f/fde9465ae6b8b824b1a8bb39503bd87a73adaeb1.jpeg

**Reply by agganitk** [reply]
ok. editing glyphs would be best.. 
anyway just for curosity.. what  GRAPHICS_OR is for...

**Reply by board707** [reply]
To be honest, I don't clearly remember :) These modes were declared in the library prototype by another author, but never completely implemented in the code.

Another illustration for font editing.  You have to edit Advance of the accent marks to move it fully before X base point of the glyph, so that they are above and below the previous letter:
![hindi_font_error](https://github.com/board707/DMD_STM32/assets/50185434/a45e146d-85d4-4dc5-a92e-eae04de6ace2)

**Reply by board707** [reply]
By the way, I will be very interested to implement the Hindi font(s) to the library, so I will be appreciated if you will share your results.

**Comment by board707** [comment]
@agganitk 
Returning to the question about GRAPHICS OR mode - in which class are you trying to do this - MonochromeSPI,   Monochrome Parallel and RGB?

**Reply by agganitk** [reply]
MonochromeSPI

**Comment by board707** [comment]
>  i need to use GRAPHICS_OR .... if yes then how to select, I dont see any function.

DMD class has a function to select graphics mode, for example
```
dmd.set_graph_mode(GRAPHICS_OR);
```
but by default only GRAPHICS_NORMAL and GRAPHICS_NOR is supported. You can try to uncomment the GRAPHICS_OR case in the switch in `DMD_MonoChrome_SPI::drawPixel()` function (DMD_MonoChrome_SPI.cpp file, line 116 and below), but since it not tested, you doing that on your own risk.
Please let me know about results if you will try to use it.

**Reply by agganitk** [reply]
GRAPHICS_OR is messing with fixed text. The fixed text also starts moving with leaving all pixel ON as text starts moving. 
Also note,
I can not run dmd.set_graph_mode(GRAPHICS_OR); I have to  edit the DMD_STMa.h  and replace GRAPHICS_NORMAL with GRAPHICS_OR.
**************************************************
uint8_t graph_mode = GRAPHICS_OR;
void set_graph_mode(uint8_t gm = GRAPHICS_OR ) {
graph_mode = gm;
}

**Comment by board707** [comment]
> The fixed text also starts moving

Yes, by default text scrollis by moving the entire screen.. To avoid this you need to add to the setup:
```
dmd.disableFastTextShift(true);
```
And to manage fixed and scrolling texts in the same screen you have to use a dual buffering mode. See the example `examples/STM32F4_RP2040/dmd_rgb_dual_buf/dmd_rgb_dual_buf.ino` for detail.

> I can not run dmd.set_graph_mode(GRAPHICS_OR);

Why you can't this? If you got a compiler errors, please show it in the message.

**Comment by agganitk** [comment]
Arduino: 1.8.5 (Windows 10), Board: "Generic STM32F103R series, STM32F103RB (20k RAM. 128k Flash), STLink, 72Mhz (Normal), Smallest (default)"

In file included from C:\Users\INTEL\Documents\Arduino\libraries\DMD_STM32-dev-V2/DMD_MonoChrome_SPI.h:11:0,

                 from C:\Users\INTEL\Documents\Arduino\dmd_mono_test\dmd_mono_test.ino:15:

C:\Users\INTEL\Documents\Arduino\libraries\DMD_STM32-dev-V2/DMD_STM32a.h: In function 'void setup()':

C:\Users\INTEL\Documents\Arduino\libraries\DMD_STM32-dev-V2/DMD_STM32a.h:326:7: error: 'void DMD::set_graph_mode(uint8)' is protected

  void set_graph_mode(uint8_t gm = GRAPHICS_NORMAL ) {

       ^

dmd_mono_test:137: error: within this context

     dmd.set_graph_mode(GRAPHICS_OR);

                                   ^

exit status 1
within this context

**Comment by board707** [comment]
> error: 'void DMD::set_graph_mode(uint8)' is protected

Understood.... The function definition should be moved to the public section of the DMD class...
